Introduce network driver

maiqueb opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR has 2 intentions:
  - move the network driver related code to a pkg located outside the virt-launcher package structure
  - move the network cache related code to a pkg located outside the virt-launcher package structure

This is required so a follow up PR can extend the existing cache mechanism to also account for the VIF caches in the bind mechanisms.

Eventually, that'll enable the load/set cache methods to be removed from the `BindMechanism` interface.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

maiqueb commented:
/cc @qinqon

dhiller reviewed (commented):
Looks good to me! One nit I have still.

EdDev commented:
/sig network

maiqueb commented:
/retest

qinqon commented on pkg/network/driver/driver_suite_test.go:
```diff
@@ -0,0 +1,13 @@
+
+func TestDriver(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Driver Suite")
```
"Network Driver Suite"

maiqueb replied:
Will do.

qinqon commented on pkg/virt-handler/vm.go:
```diff
@@ -34,6 +34,8 @@ import (
 	"sync"
 	"time"
 
+	"kubevirt.io/kubevirt/pkg/network/driver"
```
can we import this as "networkdriver" or "netdriver" since the context is "virt-handler/vm" it will give the code the missing "network" context.

maiqueb replied:
You are definitely right. Will do.

qinqon reviewed (changes requested):
Looks pretty good and small, just some nits.

maiqueb commented on pkg/network/extractors.go:
```diff
@@ -0,0 +1,48 @@
+	v1 "kubevirt.io/client-go/api/v1"
+)
+
+func GetNetworksAndCniNetworks(vmi *v1.VirtualMachineInstance) (map[string]*v1.Network, map[string]int) {
```
This section is being refactored (with something similar to what you're suggesting) on PR #5062 , and I'd prefer that to go in first. 

Both this PR & #5062 depend on #4977; once it gets in, I'll rebase this one on top of #5062, and move that network common code along as well.

I would welcome you to take a look @ the last commit in #5062, and voice your thoughts there :) 

Thanks for the input !

maiqueb commented:
This PR depends on PR #4977, and my intentions are to rebase into it #5062 as well.

As such, no point having CI looking at it.

maiqueb commented:
/retest

RamLavi commented on pkg/virt-handler/vm_test.go:
```diff
@@ -31,6 +31,8 @@ import (
 	"sync"
 	"time"
 
+	"kubevirt.io/kubevirt/pkg/network/driver"
```
should go with the kubevirt import group

RamLavi replied:
I see we have more than one kubevirt import groups.. o_O

maiqueb replied:
I'm not fixing that; moved it to all the other related stuff.

Is it OK now ?

RamLavi commented on pkg/network/cache/fake/infocache.go:
```diff
@@ -6,7 +6,7 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 
-	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/network/cache"
+	"kubevirt.io/kubevirt/pkg/network/cache"
```
what's the reasoning to put it in a different import group?

maiqueb replied:
the IDE did it, I just didn't notice.

RamLavi reviewed (changes requested):
looks good, added some import nits.

maiqueb commented:
> looks good, added some import nits.

Addressed your comments in [this forced push](https://github.com/kubevirt/kubevirt/compare/cac5b8946bdeefaff3fa33660d1426f82fb95096..7125260b864e66d79a8c11ce4f1d602f35802d48).

qinqon commented on pkg/virt-handler/vm_test.go:
```diff
@@ -31,6 +31,7 @@ import (
 	"sync"
 	"time"
 
+	network_driver "kubevirt.io/kubevirt/pkg/network/driver"
```
Remove the underscore normally at golang we just concatenate the packages and this will make it consistent with https://github.com/kubevirt/kubevirt/pull/5044/commits/162f2e94722502ae1655107118384f359c075495#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR66

maiqueb replied:
0_o

qinqon replied:
About golang naming of things https://blog.golang.org/package-names

qinqon commented on pkg/virt-launcher/virtwrap/network/network.go:
```diff
@@ -23,6 +23,7 @@ import (
 	"fmt"
 
 	v1 "kubevirt.io/client-go/api/v1"
+	network_driver "kubevirt.io/kubevirt/pkg/network/driver"
```
ditto

qinqon reviewed (changes requested):
All good just small nit on import naming

maiqueb commented:
/retest

qinqon commented:
/lgtm

qinqon commented:
/lgtm

maiqueb commented:
/retest

qinqon commented:
/lgtm

maiqueb commented:
/cc @EdDev

EdDev commented on pkg/network/cache/BUILD.bazel:
```diff
@@ -12,6 +12,7 @@ go_library(
         "//pkg/util:go_default_library",
         "//pkg/virt-launcher/virtwrap/api:go_default_library",
         "//staging/src/kubevirt.io/client-go/api/v1:go_default_library",
+        "//vendor/github.com/vishvananda/netlink:go_default_library",
```
This and some others appear in the last commit (of the error package). I guess it leaked from prev commits, right?

EdDev reviewed (commented):
Nice!

Here some naming suggestions and a note/warning for the future:
- The package alias: Can you please use the shortest possible format, e.g. `netcache`, `netdriver` and `neterrors? I helps with readability a bit.

- At the moment, the new `network` package depends on a large amount of virt-launcher sub-packages. This is a problem, as a lib is not supposed to depend inwards to the project.
This may cause issues when attempting to populate the lib with additional packages.

maiqueb commented on pkg/network/errors/generators.go:
```diff
@@ -0,0 +1,7 @@
+
+import "fmt"
+
+func CreateCriticalNetworkError(err error) *CriticalNetworkError {
```
I honestly didn't do that because `errors/errors.go` seemed awfully redundant.

But OK.

EdDev reviewed (approved):
Thanks!

maiqueb commented:
/retest

qinqon commented on pkg/network/errors/errors.go:
```diff
@@ -0,0 +1,13 @@
+package errors
```
To remove redundancy we can just move `errors.go` to `pkg/network/errors.go` since there is no sub package at `pkg/network` using it and rename it to `CriticalError` so we will have `network.CriticalError`.

Or we can even do a type alias to export unexported `networkCriticalError` as `network.CriticalError`.

But this is just cosmetics.

qinqon reviewed (approved):
All good, added one cosmetic nit, but not blocking this.

AlonaKaplan commented:
/approve thanks!

maiqueb commented:
/hold 

I'd rather have https://github.com/kubevirt/kubevirt/pull/5521 merged first, then rebase this one on top.

maiqueb commented:
/hold cancel

Well, the state CI is in, I'll take any win I can get ...
